### PR TITLE
drop keeper dir with zero copy locks

### LINKED ISSUE #58964: Some zookeeper znodes can remain after database or table are dropped
Version: 23.10.4.25 + zookeeper 3.8.2

It seems with the default replica path, if a table is created and dropped, one or more znodes associated with that table may not get deleted. 

This can be problematic if, for example, automated tests are creating and deleting unique databases and tables, causing znodes to grow. It would require manual znode cleanup from time to time, which is not ideal.

Is this intended behaviour, if so could the reasoning be explained? Is there any setting that can help with clean up of these stale znodes?

To reproduce: 

Create database:
`create database test on cluster default`

Create table:
```
CREATE TABLE IF NOT EXISTS test.mytable ON CLUSTER default
(
   a Int32,
   d DateTime64 DEFAULT now(),
) ENGINE = ReplicatedMergeTree()
PARTITION BY toYYYYMM(d)
ORDER BY (d)
```

Internally this uses the default replica path: `ENGINE = ReplicatedMergeTree('/clickhouse/tables/{uuid}/{shard}', '{replica}')`

Drop the database:
`drop database test on cluster default sync`

The `/clickhouse/tables/{uuid}` znode will not get deleted:
```
getAllChildrenNumber /clickhouse/tables
1
```

```
SELECT *
FROM system.zookeeper
WHERE path = '/clickhouse/tables'
FORMAT Vertical

Query id: 75e82258-60db-4bd6-bcae-4745023cbcaf

Row 1:
──────
name:  522b38bb-a75a-4fc3-b223-b6f0037ff78d
value:
path:  /clickhouse/tables
```

### PR DESCRIPTION
**Pull request by CheSema**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
All zero copy locks related to a table have to be dropped when the table is dropped. The directory which contains these locks has to be removed also.

Closes https://github.com/ClickHouse/ClickHouse/issues/58964

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit a5ae8f330d043f62ec1b33c25f012206e78e47cf with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57575/a5ae8f330d043f62ec1b33c25f012206e78e47cf/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7117967076">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57575/a5ae8f330d043f62ec1b33c25f012206e78e47cf/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by CheSema** [comment]
'''

2023.12.12 13:37:35.346461 [ 2015 ] {d91457b8-36e4-4178-8a26-e86c14439ee2} <Trace> test_duj5lca0.r1: Removing zookeeper lock /clickhouse/zero_copy/zero_copy_s3/0e09012d-51e4-415d-a7eb-0a0eb655da55/all_0_0_0/common_ods_ttjpicrygutbejnfhxvwucbawbasm/1 for part all_0_0_0 (files to keep: [])
2023.12.12 13:37:35.347578 [ 1304 ] {} <Trace> KeeperDispatcher: Processing requests batch, size: 1, bytes: 132
2023.12.12 13:37:35.348206 [ 2104 ] {8ceb8b4b-cf22-47d2-b3f7-b1f2bcb166a6} <Trace> test_duj5lca0.r2: Removing zookeeper lock /clickhouse/zero_copy/zero_copy_s3/0e09012d-51e4-415d-a7eb-0a0eb655da55/all_0_0_0/common_ods_ttjpicrygutbejnfhxvwucbawbasm/2 for part all_0_0_0 (files to keep: [])
2023.12.12 13:37:35.349543 [ 1304 ] {} <Trace> KeeperDispatcher: Processing requests batch, size: 1, bytes: 132
2023.12.12 13:37:35.353302 [ 2015 ] {d91457b8-36e4-4178-8a26-e86c14439ee2} <Trace> test_duj5lca0.r1: No more children left for /clickhouse/zero_copy/zero_copy_s3/0e09012d-51e4-415d-a7eb-0a0eb655da55/all_0_0_0/common_ods_ttjpicrygutbejnfhxvwucbawbasm, will try to remove the whole node
2023.12.12 13:37:35.453134 [ 1304 ] {} <Trace> KeeperDispatcher: Processing requests batch, size: 1, bytes: 130
2023.12.12 13:37:35.453765 [ 2104 ] {8ceb8b4b-cf22-47d2-b3f7-b1f2bcb166a6} <Trace> test_duj5lca0.r2: No more children left for /clickhouse/zero_copy/zero_copy_s3/0e09012d-51e4-415d-a7eb-0a0eb655da55/all_0_0_0/common_ods_ttjpicrygutbejnfhxvwucbawbasm, will try to remove the whole node
2023.12.12 13:37:35.455338 [ 1304 ] {} <Trace> KeeperDispatcher: Processing requests batch, size: 1, bytes: 130
2023.12.12 13:37:35.456430 [ 2015 ] {d91457b8-36e4-4178-8a26-e86c14439ee2} <Trace> test_duj5lca0.r1: Removed last parent zookeeper lock /clickhouse/zero_copy/zero_copy_s3/0e09012d-51e4-415d-a7eb-0a0eb655da55/all_0_0_0/common_ods_ttjpicrygutbejnfhxvwucbawbasm for part all_0_0_0 with id common_ods_ttjpicrygutbejnfhxvwucbawbasm
2023.12.12 13:37:35.467031 [ 2104 ] {8ceb8b4b-cf22-47d2-b3f7-b1f2bcb166a6} <Trace> test_duj5lca0.r2: Node with parent zookeeper lock /clickhouse/zero_copy/zero_copy_s3/0e09012d-51e4-415d-a7eb-0a0eb655da55/all_0_0_0/common_ods_ttjpicrygutbejnfhxvwucbawbasm for part all_0_0_0 with id common_ods_ttjpicrygutbejnfhxvwucbawbasm doesn't exist
'''

This is the race from tests. It has nothing with the changes, but still. 
- The tests failed, it needs to be adjusted. 
- That makes some blobs leave as abandoned on S3.

**Comment by CheSema** [comment]
something strange in ClickHouse Stateless Tests (tsan, S3 Storage). There is `Fatal` caused by `Logical Error`. But no stack trace for it.

```
2023.12.12 21:13:37.685565 [ 11090 ] {2f4806c7-42de-4a91-b3ea-d4c1ff3ade86} <Trace> test_cw8qzrvk.alter_table0 (ff6abe32-074c-455f-8dce-1d7a97cb8e7f) (Replicated OutputStream): ZooKeeperWithFaultInjection call FAILED: seed=1377394408167778856 func=tryMulti path=/clickhouse/tables/02443_detach_attach_partition_test_cw8qzrvk/alter_table/bloc
ks/all_8490796596548189825_9107552674774882560 code=No node message=Coordination error: No node, path /clickhouse/tables/02443_detach_attach_partition_test_cw8qzrvk/alter_table/block_numbers/all/block-0000000013
```
Definetly an error in response is expected.
But
```
2023.12.12 21:13:37.711823 [ 11090 ] {2f4806c7-42de-4a91-b3ea-d4c1ff3ade86} <Fatal> : Logical error: 'There is no failed OpResult'.
```

**Comment by CheSema** [comment]
test_delayed_replica_failover -- flaky test https://github.com/search?q=repo%3AClickHouse%2FClickHouse+test_delayed_replica_failover&type=issues

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -1075,6 +1073,65 @@ zkutil::ZooKeeperPtr StorageReplicatedMergeTree::getZooKeeperIfTableShutDown() c
...
+        else if (code == Coordination::Error::ZNONODE)
+        {
+            LOG_INFO(logger, "Zero copy locks directory {} is absent on ZooKeeper.", zero_copy_locks_root);
+        }
+        else if (code != Coordination::Error::ZOK)
+        {
+            LOG_ERROR(logger, "Zero copy locks were not completely removed from ZooKeeper, "
+                              "{} still exists and may contain some garbage.", zero_copy_locks_root);
+            throw zkutil::KeeperException::fromPath(code, zero_copy_locks_root);
+        }
```
Comment (tavplubix) [inline src/Storages/StorageReplicatedMergeTree.cpp:1131]: This seems to be impossible:
https://github.com/ClickHouse/ClickHouse/blob/546484d46bc3dc01b1e88404dea7170a514871cc/src/Common/ZooKeeper/ZooKeeper.h#L271-L275

Comment (tavplubix) [reply src/Storages/StorageReplicatedMergeTree.cpp:1131]: But it makes sense to write this message if `tryRemoveRecursive` returns false

**Review by tavplubix — APPROVED** [review]

File: src/Storages/StorageReplicatedMergeTree.cpp (hunk 1 above)
Comment (CheSema) [reply src/Storages/StorageReplicatedMergeTree.cpp:1131]: Do you mean that the is no other reason for `code != Coordination::Error::ZOK` after I handle `ZNOTEMPTY` and `ZNONODE` in the if-branches before?

Comment (tavplubix) [reply src/Storages/StorageReplicatedMergeTree.cpp:1131]: yes (and since you don't pass a version to `tryRemove`)

**Comment by hamidtahi** [comment]
Hi, is there an ETA for this to be merged?

**Pull request merged**